packages/test/test-end-to-end-tests/src/test/dataStoreRetrieval.spec.ts

```diff
@@ -0,0 +1,173 @@
+			public get _root() {
+				return this.root;
+			}
+
+			public get _context() {
+				return this.context;
+			}
+
+			public get _runtime() {
+				return this.runtime;
+			}
+		}
+		const innerDataObjectFactory = new DataObjectFactory(
+			"InnerDataObject",
+			InnerDataObject,
+			[],
+			[],
+		);
+
+		class OuterDataObject extends DataObject implements ITestDataObject {
+			public get _root() {
+				return this.root;
+			}
+
+			public get _context() {
+				return this.context;
+			}
+
+			public get _runtime() {
+				return this.runtime;
+			}
+
+			private readonly innerDataStoreKey = "innerDataStore";
+
+			protected async initializingFirstTime(): Promise<void> {
+				const innerDataStore = await this._context.containerRuntime.createDataStore(
+					innerDataObjectFactory.type,
+				);
+				const innerDataObject = (await innerDataStore.entryPoint?.get()) as ITestDataObject;
+				this.root.set(this.innerDataStoreKey, innerDataObject.handle);
+				await innerDataObject.handle.get();
+			}
+
+			protected async hasInitialized(): Promise<void> {
+				const innerDataStoreHandle = this.root.get<IFluidHandle<InnerDataObject>>(
+					this.innerDataStoreKey,
+				);
+				assert(innerDataStoreHandle !== undefined, "inner data store handle is missing");
+			}
+		}
+		const outerDataObjectFactory = new DataObjectFactory(
+			"OuterDataObject",
+			OuterDataObject,
+			[],
+			[],
+		);
+
+		let provider: ITestObjectProvider;
+		const innerRequestHandler = async (request: IRequest, runtime: IContainerRuntimeBase) =>
+			runtime.IFluidHandleContext.resolveHandle(request);
+
+		beforeEach(() => {
+			provider = getTestObjectProvider();
+		});
+
+		it("Requesting not visible data stores in detached container", async () => {
+			const request = provider.driver.createCreateNewRequest(provider.documentId);
+			const loader = provider.makeTestLoader();
+			const container = await loader.createDetachedContainer(provider.defaultCodeDetails);
+			// Get the default (root) dataStore from the detached container.
+			const mainDataStore = await requestFluidObject<ITestDataObject>(container, "/");
+
+			// Create another data store and make it visible by adding its handle in the root data store's DDS.
+			const dataStore2 = await mainDataStore._context.containerRuntime.createDataStore(
+				TestDataObjectType,
+			);
+			const dataObject2 = (await dataStore2.entryPoint?.get()) as ITestDataObject;
+			assert(dataObject2 !== undefined, "could not create dataStore2");
+			mainDataStore._root.set("dataStore2", dataObject2.handle);
+
+			// Request the new data store via the request API on the container.
+			const dataStore2Response = await container.request({
+				url: dataObject2.handle.absolutePath,
+			});
+			assert(
+				dataStore2Response.mimeType === "fluid/object" && dataStore2Response.status === 200,
+				"Unable to load bound data store in detached container",
+			);
+			await container.attach(request);
```

Comment:
Add a comment explaining why we're doing this.  And/or put it in an `assert.doesNotThrow(...)` call to be self-documenting